### ISSUE TITLE
Define strategy for implementing the E2E test suite for the new SceSim Editor

### ISSUE THREAD

**Issue by kbowers-ibm** [apache/incubator-kie-issues#871]
## Initial requirements:

- Easily create new tests.
- Easily understand what a test case is testing.
- Easily understand why a test broke.
- Easily run single or multiple tests locally.
- Easily debug tests that are running locally.
- Easily update tests that broke during development.
- Easily associate an issue to a regression test.
- Ability to test multiple browsers.
- Rely on tests not being flaky.
- Easily monitor the health of our test suites, looking at reports and metrics.
- Internal refactoring that don't have impact on anything visible shouldn't require tests to be changed.
- Rely on tests being as fast as possible, without sacrificing anything from the above.

## Things we'll do that will get us there:

- We will follow the example from the new test suite of the Boxed Expression Editor.
- We will have mechanisms to associate a test case with an issue.
- We will have 3 retries for each test, that will only fail if failing more than 3 times, minimizing flakyness.
- We will have PlayWright run tests on **Firefox**, **Safari** and **Chrome**, on **Windows**, **Linux**, and **macOS**.
- We will have single-line commands that anyone will be able to run on their machines for running either 
  - 1. all tests,
  - 2. a single test,
  - 3. a set of tests.
- We will have a way to debug tests during execution, placing breakpoints directly in our IDEs of choice (VS Code, Webstorm etc).
- We will have a way to look at the test playback from a CI build, or from a local run.
- We will have the tests output their results in JUnit XML format, uploading them to Buildkite on community.
- In general, we will not create giant tests. Instead, we'll create the most focused tests possible, testing one thing at a time.
- In general, big/complex models won't be created from scratch. Instead, if they're really needed they'll have a distinguishable name and very nice documentation around it. E.g., "Traffic violation example" or something like that.
- We will not use existing models from external sources. Instead, we'll define our own models either in our tests themselves or use SceSim files inside `kie-tools`.
  - models stored in `kie-tools` should follow test structuring described below.
  - for the points above, we should use probably some folders or prefix/sufix naming 

### Test types:

- SceSim features
  - E.g., A Test Scenario must have a given value and an expected value.
- Our features
    - E.g., If the same GIVEN data is shared with the multiple test scenarios, you can use the Background tab to define them only once.
- Regressions
    - E.g., `kie-issue#XXX: Lorem ipsum ....`, possiblity with a SceSim file associated with it.
- Non-functional
    - E.g., unnecessary renders, performance, etc.

> NOTE: This is not black & white, though, some tests might be in a gray area between two or more types. And more types might appear as we progress.

### Asserting strategy:
- In general, UI tests will focus on screenshots and elements being present on the screen.
- Regression tests can rely on other assertion techniques, like looking at the underlying DMN model or even the XML itself.
- Negative assertions are usually best avoided. E.g., "assert that something did not happen" or "assert that something is not present on screen"

## Remarks:
- Duplication is not always bad. Each test gets to be isolated from one another.
- Make use of well rounded primitives for reusability and readability.
- Tests being 100% detached from the source code allow for greater freedom when developing.
- Having a great set of primitives will allow other components to use them when interacting with the SceSim Editor inside their own tests. E.g., KIE Sandbox testing Dev deployments, or multi-tab editing, for example.
- Somethings are better tested as unit tests. We should be very considerate of choosing the testing strategy for something. For example, the validity of a FEEL name should be tested by unit tests, but the fact that the UI renders invalid names in red, should be an E2E test.

## Things we won't do:
- Mock too much to simulate tests that should actually be somewhere else: E.g., multi-file editing on KIE Sandbox / VS Code Extension
- Test running the DMNs (that will be a separate test suite..., not associated with any component in particular.. like `@kie-tools/dmn-e2e-tests`, or even in another repo!)

---

## Action plan:

1. **[IN PROGRESS]** Exploratory phase: phew test cases, predefined, will be used to create some ground work.
    - Creating the StoryBook and migrating the Dev Webapp there.
    - Defining basic tests for dragging nodes into the diagram and making connections between them.
2. Expansion phase: create more tests to identify problems at a larger scale.
    - Refine our primitives, refactor the first tests, get a better sense of how things are going to look like once the test suite is bigger
    - Start integrating with the CI, improving documentation etc.
3. Maturity phase:
    - Finish adding the initial tests we want, improve pain points and smooth edges out.
    - Ask for feedback, share the progress.
    - Document how other people on the team can use and interact with the test suites.

## Next steps:
    - Monitor the quality of our tests, see what is working well and what is not.
    - Keep an eye on PRs to make sure everyone is adding tests when changing things.
    - Help propagate the testing culture to the team, offering help and acting as owners of the test suites.
    - Continuous work. It doesn't stop here.
